### PR TITLE
feat(sdk-coin-atom): add staking transaction builders

### PR DESCRIPTION
**Pull request by dpkjnr**
Ticket: BG-69278

BREAKING CHANGE: type MessageData is changed

### CODE REVIEW

File: modules/sdk-coin-atom/src/lib/iface.ts [hunk 1]
```diff
@@ -24,15 +37,15 @@ export interface FeeData {
  */
 export interface TxData extends AtomTransaction {
   id?: string;
+  type?: TransactionType;
```
Comment (margueriteblair) [inline modules/sdk-coin-atom/src/lib/iface.ts:40]: is there a reason id and type are both optional? they both seem important

File: modules/sdk-coin-atom/test/unit/transactionBuilder/StakingActivateBuilder.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import { getBuilderFactory } from '../getBuilderFactory';
+import * as testData from '../../resources/atom';
+import should from 'should';
+import { toHex, TransactionType } from '@bitgo/sdk-core';
+import { Transaction as AtomTransaction } from '../../../src/lib/transaction';
+import { fromBase64 } from '@cosmjs/encoding';
+
+describe('Atom Delegate txn Builder', () => {
+  const factory = getBuilderFactory('tatom');
```
[the remark this answers is not in the document]
Comment (margueriteblair) [reply modules/sdk-coin-atom/test/unit/transactionBuilder/StakingActivateBuilder.ts:9]: Like here: https://github.com/BitGo/BitGoJS/blob/master/modules/sdk-coin-ada/test/unit/StakingActivateBuilder.ts#L77 import axios and submit to a fullnode if atom has those resources availble to test

File: modules/sdk-coin-atom/test/unit/transactionBuilder/StakingDeactivateBuilder.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import { getBuilderFactory } from '../getBuilderFactory';
+import * as testData from '../../resources/atom';
+import should from 'should';
+import { toHex, TransactionType } from '@bitgo/sdk-core';
+import { Transaction as AtomTransaction } from '../../../src/lib/transaction';
+import { fromBase64 } from '@cosmjs/encoding';
+
+describe('Atom Undelegate txn Builder', () => {
```
Comment (margueriteblair) [inline modules/sdk-coin-atom/test/unit/transactionBuilder/StakingDeactivateBuilder.ts:8]: Same for undelegate, if you can make an integration test, that would be great

File: modules/sdk-coin-atom/test/unit/transactionBuilder/StakingActivateBuilder.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
+import { getBuilderFactory } from '../getBuilderFactory';
+import * as testData from '../../resources/atom';
+import should from 'should';
+import { toHex, TransactionType } from '@bitgo/sdk-core';
+import { Transaction as AtomTransaction } from '../../../src/lib/transaction';
+import { fromBase64 } from '@cosmjs/encoding';
+
+describe('Atom Delegate txn Builder', () => {
```
Comment (ewangbitgo) [inline modules/sdk-coin-atom/test/unit/transactionBuilder/StakingActivateBuilder.ts:8]: please add more test, include "deserialize, add signature", see sui's test as a reference

File: modules/sdk-coin-atom/src/lib/iface.ts (hunk 1 above)
Comment (dpkjnr) [reply modules/sdk-coin-atom/src/lib/iface.ts:40]: both are optional in the baseTransaction class

File: modules/sdk-coin-atom/src/atom.ts [hunk 5]
```diff
@@ -158,7 +157,32 @@ export class Atom extends BaseCoin {
...
+    };
+  }
+
+  /** @inheritDoc **/
+  signTransaction(params: SignTransactionOptions): Promise<SignedTransaction> {
+    throw new Error('Method not implemented.');
+  }
+
+  /** @inheritDoc **/
+  async isWalletAddress(params: VerifyAddressOptions): Promise<boolean> {
```
[the remark this answers is not in the document]
Comment (dpkjnr) [reply modules/sdk-coin-atom/src/atom.ts:185]: I didn't see this implemented for recently onboarded coins. Not sure if it is even used

File: modules/sdk-coin-atom/test/unit/transactionBuilder/StakingActivateBuilder.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (dpkjnr) [reply modules/sdk-coin-atom/test/unit/transactionBuilder/StakingActivateBuilder.ts:12]: done

File: modules/sdk-coin-atom/test/unit/transactionBuilder/StakingDeactivateBuilder.ts (hunk 3 above)
Comment (dpkjnr) [reply modules/sdk-coin-atom/test/unit/transactionBuilder/StakingDeactivateBuilder.ts:8]: done

File: modules/sdk-coin-atom/test/unit/transactionBuilder/StakingActivateBuilder.ts (hunk 4 above)
Comment (dpkjnr) [reply modules/sdk-coin-atom/test/unit/transactionBuilder/StakingActivateBuilder.ts:11]: Added more tests and verifying more data from the result

File: modules/sdk-coin-atom/test/unit/transactionBuilder/StakingActivateBuilder.ts [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    const serializedTx = tx.toBroadcastFormat();
+
+    try {
+      const res = await axios.post(
+        'https://rest.sentry-01.theta-testnet.polypore.xyz/cosmos/tx/v1beta1/txs',
+        {
+          mode: 'BROADCAST_MODE_BLOCK',
+          tx_bytes: serializedTx,
+        },
+        axiosConfig
```
Comment (margueriteblair) [inline modules/sdk-coin-atom/test/unit/transactionBuilder/StakingActivateBuilder.ts:91]: and to confirm, when you submitted this, it was successful on chain?

Comment (dpkjnr) [reply modules/sdk-coin-atom/test/unit/transactionBuilder/StakingActivateBuilder.ts:91]: yes it was. https://explorer.theta-testnet.polypore.xyz/transactions/3D9583366D1D1AA82229800DE323F6B26E0170AD4CE610D19F08CD45D36505E2 this is the one I guess. Anyone can run it, only the sequence number needs to be changed

File: modules/sdk-coin-atom/package.json [hunk 7]
```diff
@@ -54,6 +55,7 @@
   "devDependencies": {
     "@bitgo/sdk-api": "^1.9.10",
     "@bitgo/sdk-test": "^1.2.15",
-    "@types/lodash": "^4.14.183"
+    "@types/lodash": "^4.14.183",
+    "axios": "^1.3.4"
```
Comment (sachushaji) [inline modules/sdk-coin-atom/package.json:59]: No yarn lock updates for this change?

**Review by mukeshsp — APPROVED** [review]
LGTM

File: modules/sdk-coin-atom/package.json (hunk 7 above)
Comment (dpkjnr) [reply modules/sdk-coin-atom/package.json:59]: added

**Review by sachushaji — APPROVED** [review]
package.json and yarn.lock changes LGTM

**Pull request merged**